csr/csr.go

```diff
@@ -0,0 +1,85 @@
+package csr
+
+import (
+	"crypto"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/letsencrypt/boulder/core"
+)
+
+// maxCNLength is the maximum length allowed for the common name as specified in RFC 5280
+const maxCNLength = 64
+
+// This map is used to detect algorithms in crypto/x509 that
+// are no longer considered sufficiently strong.
+// * No MD2, MD5, or SHA-1
+// * No DSA
+//
+// SHA1WithRSA is allowed because there's still a fair bit of it
+// out there, but we should try to remove it soon.
+var badSignatureAlgorithms = map[x509.SignatureAlgorithm]bool{
+	x509.UnknownSignatureAlgorithm: true,
+	x509.MD2WithRSA:                true,
+	x509.MD5WithRSA:                true,
+	x509.DSAWithSHA1:               true,
+	x509.DSAWithSHA256:             true,
+	x509.ECDSAWithSHA1:             true,
+}
+
+// VerifyCSR checks the validity of a x509.CertificateRequest
+func VerifyCSR(csr *x509.CertificateRequest, maxNames int, keyPolicy *core.KeyPolicy, pa core.PolicyAuthority, regID int64) error {
+	key, ok := csr.PublicKey.(crypto.PublicKey)
+	if !ok {
+		return errors.New("invalid public key in CSR")
+	}
+	if err := keyPolicy.GoodKey(key); err != nil {
+		return fmt.Errorf("invalid public key in CSR: %s", err)
+	}
+	if badSignatureAlgorithms[csr.SignatureAlgorithm] {
+		// go1.6 provides a stringer for x509.SignatureAlgorithm but 1.5.x
+		// does not
+		return errors.New("signature algorithm not supported")
+	}
+	if err := csr.CheckSignature(); err != nil {
+		return errors.New("invalid signature on CSR")
+	}
+	if len(csr.DNSNames) == 0 && csr.Subject.CommonName == "" {
+		return errors.New("at least one DNS name is required")
+	}
+	if len(csr.Subject.CommonName) > maxCNLength {
+		return fmt.Errorf("CN was longer than %d bytes", maxCNLength)
+	}
+	if maxNames > 0 && len(csr.DNSNames) > maxNames {
```

Comment:
This means that this function has a required precondition that NormalizeCSR be called first, or this would allow 101 names (including CN). If we want to specify that precondition, we should document it, but we could also just call it at the top of this function.